Framework/Crystal/src/ConvertPeaksWorkspace.cpp

```diff
@@ -0,0 +1,164 @@
+// Mantid Repository : https://github.com/mantidproject/mantid
+//
+// Copyright &copy; 2021 ISIS Rutherford Appleton Laboratory UKRI,
+//   NScD Oak Ridge National Laboratory, European Spallation Source,
+//   Institut Laue - Langevin & CSNS, Institute of High Energy Physics, CAS
+// SPDX - License - Identifier: GPL - 3.0 +
+
+#include "MantidCrystal/ConvertPeaksWorkspace.h"
+#include "MantidAPI/MatrixWorkspace.h"
+#include "MantidAPI/Run.h"
+#include "MantidAPI/Sample.h"
+#include "MantidDataObjects/LeanElasticPeak.h"
+#include "MantidDataObjects/LeanElasticPeaksWorkspace.h"
+#include "MantidDataObjects/PeaksWorkspace.h"
+#include "MantidGeometry/Crystal/OrientedLattice.h"
+#include "MantidGeometry/Instrument/Goniometer.h"
+#include "MantidKernel/Logger.h"
+
+namespace Mantid {
+namespace Crystal {
+
+using namespace Mantid::API;
+using namespace Mantid::DataObjects;
+using namespace Mantid::Geometry;
+using namespace Mantid::Kernel;
+
+/// Config logger
+namespace {
+Logger logger("ConvertPeaksWorkspace");
+}
+
+DECLARE_ALGORITHM(ConvertPeaksWorkspace)
+
+/**
+ * @brief initialization of alg
+ *
+ */
+void ConvertPeaksWorkspace::init() {
+  // Input peakworkspace
+  declareProperty(std::make_unique<WorkspaceProperty<IPeaksWorkspace>>("PeakWorkspace", "", Direction::Input),
+                  "Workspace of Indexed Peaks");
+
+  // donor workspace if going from lean to regular
+  declareProperty(std::make_unique<WorkspaceProperty<Workspace>>("InstrumentWorkspace", "", Direction::Input,
+                                                                 PropertyMode::Optional),
+                  "Donor PeaksWorkspace with instrument for conversion");
```

Comment:
You will need to add validation for this.

This will segfault
```python
pw = CreatePeaksWorkspace(OutputType='LeanElasticPeak', NumberOfPeaks=1)
md = CreateMDWorkspace(Extents='-1,1', Names='A', Units='A')
pw2 = ConvertPeaksWorkspace(pw, md)
```

So can you add something like this to `validateInputs`

```c++
Workspace_sptr workspace = getProperty("InstrumentWorkspace");
ExperimentInfo_sptr inputExperimentInfo = std::dynamic_pointer_cast<ExperimentInfo>(workspace);
if (!inputExperimentInfo)
  issues["InstrumentWorkspace"] = ...
```

Could you also update the property description, so just Workspace and not PeaksWorkspace.

Reply:
@rosswhitfield The description is updated, and a new validation case is added to check for invalid instrument as suggested.